server/server/src/metrics.rs

```diff
@@ -0,0 +1,221 @@
+use crate::keys::public_key;
+use crate::{keys, ServerError, ServerState};
+use lazy_static::lazy_static;
+use libsecp256k1::PublicKey;
+use lockbook_models::api::FileUsage;
+use lockbook_models::file_metadata::{EncryptedFileMetadata, FileType};
+use log::error;
+use prometheus::{register_int_gauge_vec, IntGaugeVec};
+use prometheus_static_metric::make_static_metric;
+use redis::{AsyncCommands, AsyncIter};
+use redis_utils::converters::JsonGet;
+use std::collections::HashSet;
+use std::fmt::Debug;
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+use uuid::Uuid;
+
+make_static_metric! {
+    pub struct MetricsStatistics: IntGauge {
+        "type" => {
+            total_users,
+            active_users,
+            total_documents,
+            total_document_bytes,
+        },
+    }
+}
+
+lazy_static! {
+    pub static ref METRICS_COUNTERS_VEC: IntGaugeVec = register_int_gauge_vec!(
+        "lockbook_metrics_counters",
+        "Lockbook's basic metrics of users and files derived from redis.",
+        &["type"]
+    )
+    .unwrap();
+    pub static ref METRICS_STATISTICS: MetricsStatistics =
+        MetricsStatistics::from(&METRICS_COUNTERS_VEC);
+}
+
+// # required metrics
+// X total users
+// active percent (just active users / total users in metrics.lockbook.com)
+// active users
+// new users
+// X total documents
+// X total bytes
+// premium users and percentage
+
+#[derive(Debug)]
+pub enum MetricsError {}
+
+// pub struct MetricsPreProcessingInfo<'a> {
+//     pub public_keys: Vec<PublicKey>,
+//     pub ids: HashMap<&'a PublicKey, Uuid>,
+//     pub metadatas: HashMap<&'a PublicKey, EncryptedFileMetadata>
+// }
+
+pub const TWO_DAYS_IN_MILLIS: u128 = 172800000;
+
+pub async fn init(server_state: &Arc<ServerState>) {
+    let state_clone = server_state.clone();
+
+    tokio::spawn(async move {
+        if let Err(e) = start(state_clone).await {
+            error!("Metrics error: {:?}", e)
+        }
+    });
+}
+
+pub async fn start(server_state: Arc<ServerState>) -> Result<(), ServerError<MetricsError>> {
+    loop {
+        let public_keys = get_all_public_keys(&server_state).await?;
+        METRICS_STATISTICS.total_users.set(public_keys.len() as i64);
+
+        let all_file_ids = get_owned_files_id(&server_state, &public_keys).await?;
+        METRICS_STATISTICS.total_documents.set(
+            all_file_ids
+                .iter()
+                .map(|users_files| users_files.len())
+                .sum::<usize>() as i64,
+        );
+
+        let metadatas = get_file_metadata(&server_state, &all_file_ids).await?;
+        let total_bytes = get_total_size(&server_state, &metadatas).await?;
```

Comment:
At this moment you will have used a massive amount of memory, you'll have all metas for all users in ram.

If our server is at a healthy 10/16gb at this moment, then this will cause our server to [OOM](https://en.wikipedia.org/wiki/Out_of_memory) and crash.

You should re-write this with the following constraints:
+ Don't keep more than the following in memory at once:
    + All public keys
    + All needed memory for at most 1 user

+ Additionally you should sprinkle some sleeps in between operations like this:

```
for key in public_keys_keys {
        let public_key = con.maybe_json_get(&key).await?.ok_or_else(|| {
            internal!(
                "Cannot retrieve public_key despite having a valid key: {:?}",
                key
            )
        })?;

        public_keys.push(public_key)
```

If we tried to compute metrics as quickly as possible we'd put considerable load on redis (think about `MONITOR` and all the calls just being metric calls for like a few seconds.

Thankfully we don't need this to be fast, slower is better.